From d9a4863535261cba4ef992ec065b94b2775c9268 Mon Sep 17 00:00:00 2001
From: Johannes Demel <jdemel@gnuradio.org>
Date: Fri, 4 Jul 2025 16:17:57 +0200
Subject: [PATCH 03/12] ci: Remove obsolete CI, add new CI

This PR switches to later Ubuntu versions, and runs the CI with later
compilers.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>
---
 .github/workflows/run-tests.yml | 20 ++++----------------
 CMakeLists.txt                  | 15 +++++++++++----
 apps/CMakeLists.txt             |  2 ++
 cmake/Modules/VolkAddTest.cmake |  1 +
 lib/CMakeLists.txt              |  1 +
 5 files changed, 19 insertions(+), 20 deletions(-)

diff --git a/.github/workflows/run-tests.yml b/.github/workflows/run-tests.yml
index 8c2b5e4..7e87e35 100644
--- a/.github/workflows/run-tests.yml
+++ b/.github/workflows/run-tests.yml
@@ -18,10 +18,6 @@ jobs:
       fail-fast: false
       matrix:
         compiler:
-          - { name: g++-9, cc: gcc-9, cxx: g++-9, distro: ubuntu-20.04 }
-          - { name: g++-9, cc: gcc-9, cxx: g++-9, distro: ubuntu-22.04-arm }
-          - { name: g++-10, cc: gcc-10, cxx: g++-10, distro: ubuntu-20.04 }
-          - { name: g++-10, cc: gcc-10, cxx: g++-10, distro: ubuntu-22.04-arm }
           - { name: g++-11, cc: gcc-11, cxx: g++-11, distro: ubuntu-22.04 }
           - { name: g++-11, cc: gcc-11, cxx: g++-11, distro: ubuntu-22.04-arm }
           - { name: g++-12, cc: gcc-12, cxx: g++-12, distro: ubuntu-22.04 }
@@ -30,15 +26,8 @@ jobs:
           - { name: g++-13, cc: gcc-13, cxx: g++-13, distro: ubuntu-24.04-arm }
           - { name: g++-14, cc: gcc-14, cxx: g++-14, distro: ubuntu-24.04 }
           - { name: g++-14, cc: gcc-14, cxx: g++-14, distro: ubuntu-24.04-arm }
-          - { name: clang-10, cc: clang-10, cxx: clang++-10, distro: ubuntu-20.04 }
-          - { name: clang-11, cc: clang-11, cxx: clang++-11, distro: ubuntu-20.04 }
-          - { name: clang-11, cc: clang-11, cxx: clang++-11, distro: ubuntu-22.04-arm }
-          - { name: clang-12, cc: clang-12, cxx: clang++-12, distro: ubuntu-22.04 }
-          - { name: clang-12, cc: clang-12, cxx: clang++-12, distro: ubuntu-22.04-arm }
-          - { name: clang-13, cc: clang-13, cxx: clang++-13, distro: ubuntu-22.04 }
-          - { name: clang-13, cc: clang-13, cxx: clang++-13, distro: ubuntu-22.04-arm }
           - { name: clang-14, cc: clang-14, cxx: clang++-14, distro: ubuntu-22.04 }
-          - { name: clang-14, cc: clang-14, cxx: clang++-14, distro: ubuntu-22.04-arm }
+          # - { name: clang-14, cc: clang-14, cxx: clang++-14, distro: ubuntu-22.04-arm } # possibly broken runner: https://github.com/actions/runner-images/issues/8659
           - { name: clang-15, cc: clang-15, cxx: clang++-15, distro: ubuntu-22.04 }
           - { name: clang-15, cc: clang-15, cxx: clang++-15, distro: ubuntu-22.04-arm }
           - { name: clang-16, cc: clang-16, cxx: clang++-16, distro: ubuntu-24.04 }
@@ -47,6 +36,8 @@ jobs:
           - { name: clang-17, cc: clang-17, cxx: clang++-17, distro: ubuntu-24.04-arm }
           - { name: clang-18, cc: clang-18, cxx: clang++-18, distro: ubuntu-24.04 }
           - { name: clang-18, cc: clang-18, cxx: clang++-18, distro: ubuntu-24.04-arm }
+          - { name: clang-19, cc: clang-19, cxx: clang++-19, distro: ubuntu-24.04 }
+          - { name: clang-19, cc: clang-19, cxx: clang++-19, distro: ubuntu-24.04-arm }
 
     runs-on: ${{ matrix.compiler.distro }}
 
@@ -89,9 +80,6 @@ jobs:
       fail-fast: false
       matrix:
         include:
-          - arch: aarch64
-            distro: ubuntu20.04
-            compiler: { name: clang-9, cc: clang-9, cxx: clang++-9 }
           - arch: armv7
             distro: ubuntu22.04
             compiler: { name: g++, cc: gcc, cxx: g++ }
@@ -110,7 +98,7 @@ jobs:
       - uses: actions/checkout@v4
         with:
           submodules: "recursive"
-      - uses: uraimo/run-on-arch-action@v2
+      - uses: uraimo/run-on-arch-action@v3
         name: Build in non-x86 container
         id: build
         with:
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5afb2dc..dffd8ef 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -11,7 +11,7 @@
 # Project setup
 ########################################################################
 # We use `IS_64BIT now: https://cmake.org/cmake/help/latest/command/cmake_host_system_information.html
-cmake_minimum_required(VERSION 3.10)
+cmake_minimum_required(VERSION 3.22)
 
 set(CMAKE_BUILD_TYPE
     ${CMAKE_BUILD_TYPE}
@@ -21,9 +21,6 @@ project(volk)
 enable_language(CXX)
 enable_language(C)
 
-set(CMAKE_C_STANDARD 11)
-set(CMAKE_CXX_STANDARD 17)
-
 enable_testing()
 
 ########################################################################
@@ -35,12 +32,22 @@ include(CheckCXXCompilerFlag)
 check_cxx_compiler_flag(-fcx-limited-range HAVE_CX_LIMITED_RANGE)
 if(HAVE_CX_LIMITED_RANGE)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fcx-limited-range")
+    check_cxx_compiler_flag(-Wno-unused-command-line-argument
+                            HAVE_WNO_UNUSED_CMD_LINE_ARG)
+    if(HAVE_WNO_UNUSED_CMD_LINE_ARG)
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-command-line-argument")
+    endif(HAVE_WNO_UNUSED_CMD_LINE_ARG)
 endif(HAVE_CX_LIMITED_RANGE)
 
 include(CheckCCompilerFlag)
 check_c_compiler_flag(-fcx-limited-range HAVE_C_LIMITED_RANGE)
 if(HAVE_C_LIMITED_RANGE)
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fcx-limited-range")
+    check_cxx_compiler_flag(-Wno-unused-command-line-argument
+                            HAVE_WNO_UNUSED_CMD_LINE_ARG)
+    if(HAVE_WNO_UNUSED_CMD_LINE_ARG)
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-command-line-argument")
+    endif(HAVE_WNO_UNUSED_CMD_LINE_ARG)
 endif(HAVE_C_LIMITED_RANGE)
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
diff --git a/apps/CMakeLists.txt b/apps/CMakeLists.txt
index 98d002a..34d6e79 100644
--- a/apps/CMakeLists.txt
+++ b/apps/CMakeLists.txt
@@ -21,6 +21,7 @@ add_executable(
     volk_profile
     ${CMAKE_CURRENT_SOURCE_DIR}/volk_profile.cc ${PROJECT_SOURCE_DIR}/lib/qa_utils.cc
     ${CMAKE_CURRENT_SOURCE_DIR}/volk_option_helpers.cc)
+target_compile_features(volk_profile PUBLIC cxx_std_17)
 
 if(MSVC)
     target_include_directories(
@@ -57,6 +58,7 @@ install(
 # MAKE volk-config-info
 add_executable(volk-config-info volk-config-info.cc
                                 ${CMAKE_CURRENT_SOURCE_DIR}/volk_option_helpers.cc)
+target_compile_features(volk-config-info PUBLIC cxx_std_17)
 
 if(ENABLE_STATIC_LIBS)
     target_link_libraries(volk-config-info volk_static)
diff --git a/cmake/Modules/VolkAddTest.cmake b/cmake/Modules/VolkAddTest.cmake
index 8104919..f4aef6e 100644
--- a/cmake/Modules/VolkAddTest.cmake
+++ b/cmake/Modules/VolkAddTest.cmake
@@ -24,6 +24,7 @@ function(VOLK_GEN_TEST executable_name)
                           "SOURCES;TARGET_DEPS;EXTRA_LIB_DIRS;ENVIRONS;ARGS" ${ARGN})
     add_executable(${executable_name} ${VOLK_TEST_SOURCES})
     target_link_libraries(${executable_name} ${VOLK_TEST_TARGET_DEPS})
+    target_compile_features(${executable_name} PUBLIC cxx_std_17)
 endfunction()
 
 ########################################################################
diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index 588db44..db3e78d 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -564,6 +564,7 @@ if(VOLK_CPU_FEATURES)
         PRIVATE $<TARGET_PROPERTY:CpuFeatures::cpu_features,INTERFACE_INCLUDE_DIRECTORIES>
     )
 endif()
+target_compile_features(volk_obj PUBLIC c_std_17)
 
 #Configure object target properties
 if(NOT MSVC)
-- 
2.47.2

